bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthConnector.java

```diff
@@ -0,0 +1,341 @@
+ *
+ * @author Michael Bock - Initial contribution
+ * @author Gary Tse - ESH adaptation
+ *
+ */
+public class OAuthConnector {
+
+    @NonNull
+    private static final String HTTP_CLIENT_CONSUMER_NAME = "OAuthConnector";
+
+    private final HttpClientFactory httpClientFactory;
+
+    private static final Logger logger = LoggerFactory.getLogger(OAuthConnector.class);
+    private final Gson gson;
+    // private SslService sslService;
+
+    public OAuthConnector(HttpClientFactory httpClientFactory) {
+        this.httpClientFactory = httpClientFactory;
+        gson = new GsonBuilder().setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES).create();
+    }
+
+    /**
+     * Authorization Code Grant
+     *
+     * @param authorizationEndpoint The end point of the authorization provider that performs authorization of the
+     *            resource owner
+     * @param clientId Client identifier (will be URL-encoded)
+     * @param redirectURI RFC 6749 section 3.1.2 (will be URL-encoded)
+     * @param state Recommended to enhance security (will be URL-encoded)
+     * @param scope Optional space separated list of scope (will be URL-encoded)
+     *
+     * @return A URL based on the authorizationEndpoint, with query parameters added.
+     * @see https://tools.ietf.org/html/rfc6749#section-4.1.1
+     */
+    public static String getAuthorizationUrl(@NonNull String authorizationEndpoint, @NonNull String clientId,
+            @Nullable String redirectURI, @Nullable String state, @Nullable String scope) {
+        StringBuilder authorizationUrl = new StringBuilder(authorizationEndpoint);
+
+        if (authorizationUrl.indexOf("?") == -1) {
+            authorizationUrl.append('?');
+        } else {
+            authorizationUrl.append('&');
+        }
+
+        try {
+            authorizationUrl.append("response_type=code");
+
+            authorizationUrl.append("&client_id=").append(URLEncoder.encode(clientId, StandardCharsets.UTF_8.name()));
+
+            if (state != null) {
+                authorizationUrl.append("&state=").append(URLEncoder.encode(state, StandardCharsets.UTF_8.name()));
+            }
+            if (redirectURI != null) {
+                authorizationUrl.append("&redirect_uri=")
+                        .append(URLEncoder.encode(redirectURI, StandardCharsets.UTF_8.name()));
+            }
+
+            if (scope != null) {
+                authorizationUrl.append("&scope=").append(URLEncoder.encode(scope, StandardCharsets.UTF_8.name()));
+            }
+
+        } catch (UnsupportedEncodingException e) {
+            // never happens
+            logger.error("Unknown encoding {}", e.getMessage(), e);
+        }
+        return authorizationUrl.toString();
+    }
+
+    /**
+     * Resource Owner Password Credentials Grant
+     *
+     * @see https://tools.ietf.org/html/rfc6749#section-4.3
+     *
+     * @param tokenUrl
```

Comment:
if you feel param or exception descriptions would be redundant, you can remove the following lines